Add documentation for installing and using docker with NILRT (#54)
ni/nilrt-docs: docs/source/docker/docker.rst

==================================================
NI Linux Real-Time and Docker
==================================================

.. contents:: Table of Contents
   :depth: 2
   :local: 

Introduction
============

The `Docker platform`_ allows for developers to run applications in consistent environments across

> multiple hardware architectures and operating systems through the use of containers.

Reviewer: I'm thinking this sentence is incorrect: Docker does not *per se* handle cross-architecture concerns, and it is usually[\*][1] rather coupled to Linux kernel features.

Perhaps we can say it helps with producing consistent environments and isolating the processes within them, and just leave it at that. Feel free to reply (or chat out-of-band) if you think more context is warranted.

[1]: <https://docs.docker.com/desktop/windows/permission-requirements/#windows-containers>

Docker can be used on NI Linux Real-Time with NI hardware for test and measurement use cases.
The document encompasses the basics of how to install Docker and run containers from an
NI Linux Real-Time system.

.. _Docker platform: https://www.docker.com

Installation
============

To install Docker on an NI Linux Real-Time system, simply run the following command
from your target:

.. code:: bash
    
    opkg install docker-ce

.. note::

>     Currently Docker will not function correctly if you have LabVIEW RT installed.

Reviewer 2: Is there a workaround for this that we should list? Or is the only workaround to not install LV RT?

Reviewer: From what I understand, we'll need LabVIEW to start using cgroups v2 (see AB#2326685 and associated wiki page). I don't recall if we have a workaround that gets both working simultaneously.

Author: If I'm not mistaken a work around is currently in development, but it's unclear when it will be complete, so it was determined that we put this note in for now, and it will be removed when we have them working together

Reviewer 2: That makes sense to me. I think we can leave this as is, then.

==================================================
NI Linux Real-Time and Docker
==================================================

.. contents:: Table of Contents
   :depth: 2
   :local: 

Introduction
============

> The docker platform allows for developers to run applications in consistent environments across

Reviewer: Apologies for being unclear: by "in prose" I meant "in all text throughout this document except command-line invocations and URLs."

hardware and across operating systems.
It is possible to utilize the docker platform on NI Linux Real-Time for containers to be used on
NI hardware in a variety of test and measurement situations.
The document encompasses the basics of how to install docker and run containers from an

> NI Linux Real-Time system.

Reviewer: Missed an underscore in v2:

    `Docker platform`

should be

    `Docker Platform`_

    Status: Downloaded newer image for hello-world:latest

    Hello from Docker!
    This message shows that your installation appears to be working correctly.

    To generate this message, Docker took the following steps:
     1. The Docker client contacted the Docker daemon.
     2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
        (amd64)
     3. The Docker daemon created a new container from that image which runs the
        executable that produces the output you are currently reading.
     4. The Docker daemon streamed that output to the Docker client, which sent it
        to your terminal.

    To try something more ambitious, you can run an Ubuntu container with:
     $ docker run -it ubuntu bash

    Share images, automate workflows, and more with a free Docker ID:
     https://hub.docker.com/

    For more examples and ideas, visit:
     https://docs.docker.com/get-started/

Starting Containers with Host Networking
-----------------

There are a variety of situations that call for containers to have internet access.
By default, containers on NI Linux Real-Time will only used bridged adapters, not the host network,

> but using the host network is sometimes required.

Author: I reworded it to point out that the need for host networking is because we don't currently support networking via the default bridged adapter

==================================================
NI Linux Real-Time and Docker
==================================================

.. contents:: Table of Contents
   :depth: 2
   :local: 

Introduction
============

The `Docker platform`_ allows for developers to run applications in consistent environments across
hardware and operating systems.
Docker can be used on NI Linux Real-Time with NI hardware for test and measurement use cases.
The document encompasses the basics of how to install Docker and run containers from an
NI Linux Real-Time system.

.. _Docker platform: https://www.docker.com

Installation
============

To install Docker on an NI Linux Real-Time system, simply run the following command
from your target:

.. code:: bash
    
    opkg install docker-ce

> NOTE: Currently Docker will not function correctly if you have LabVIEW RT installed

Reviewer: Can use the ["note" admonition][1] for this.

```suggestion
.. note::

    Currently Docker will not function correctly if you have LabVIEW RT installed.
```

[1]: <https://docutils.sourceforge.io/docs/ref/rst/directives.html#specific-admonitions>

    Digest: sha256:88ec0acaa3ec199d3b7eaf73588f4518c25f9d34f58ce9a0df68429c5af48e8d
    Status: Downloaded newer image for hello-world:latest

    Hello from Docker!
    This message shows that your installation appears to be working correctly.

    To generate this message, Docker took the following steps:
     1. The Docker client contacted the Docker daemon.
     2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
        (amd64)
     3. The Docker daemon created a new container from that image which runs the
        executable that produces the output you are currently reading.
     4. The Docker daemon streamed that output to the Docker client, which sent it
        to your terminal.

    To try something more ambitious, you can run an Ubuntu container with:
     $ docker run -it ubuntu bash

    Share images, automate workflows, and more with a free Docker ID:
     https://hub.docker.com/

    For more examples and ideas, visit:
     https://docs.docker.com/get-started/

Starting Containers with Host Networking
========================================

By default, Docker containers will use bridged adapters, but this is not currently supported by

> NI Linux Real-Time.

Reviewer: Since the default doesn't work, should we also mention that [networking can be disabled][1]?

[1]: <https://docs.docker.com/network/drivers/none/>

> Thus, one must use host networking if they want access to the internet from their container.

Reviewer: Nit: not just the Internet, but any of the host's networking.

```suggestion
Thus, one must use host networking if they want access to the network from their container.
```

> 

Reviewer: Should provide a link to Docker's docs and mention that this reduces isolation for readers who won't follow it.

```suggestion
See Docker's documentation on `host networking`_ for details.

.. _host networking: https://docs.docker.com/network/drivers/host/

.. warning::

    Host networking reduces the isolation of the container.

```

> Host networking can be forced by starting Docker with the tags `--network=host` or `--network host`.

Reviewer: Reword for accuracy and use [double-backtick for literal text][1].

```suggestion
To use host networking, create the Docker container with ``--network=host`` or ``--network host`` arguments.
```

[1]: <https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#inline-literals>
